File: Leetcode/872_Leaf-Similar_Trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function (root1, root2) {
    let result1 = [];
    let result2 = [];

    traversal(root1, result1);
    traversal(root2, result2);

    if (result1.length !== result2.length) {
        return false;
    }

    for (let i = 0; i < result1.length; i++) {
        if (result1[i] !== result2[i]) {
            return false;
        }
    }

    return true;
};

const traversal = (root, result) => {
    if (root) {
        traversal(root.left, result);
        if (root.left === null && root.right === null) {
            result.push(root.val);
        }
        traversal(root.right, result);
    }
}

File: Leetcode/0872/872_Leaf-Similar_Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> leaf1;
    vector<int> leaf2;
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
    
        addLeaves(leaf1, root1);
        addLeaves(leaf2, root2);
        
        if(leaf1.size() != leaf2.size())
        {
            return false;
        }
        
        for(unsigned int i = 0; i < leaf1.size(); i++)
        {
            if(leaf1[i] != leaf2[i])
            {
                return false;
            }
        }
        return true;
    }
    
    void addLeaves(vector<int>& leaf, TreeNode* root)
    {
        if(root == nullptr)
        {
            return;
        }
        addLeaves(leaf, root->left);
        if(root->left == nullptr && root->right == nullptr)
        {
            leaf.push_back(root->val);
        }
        addLeaves(leaf, root->right);
    }
};